lib/Ssl/SslInterface.cpp

```diff
@@ -316,6 +319,62 @@ int sslRand(int32_t* value) {
   return 0;
 }
 
+
+int rsaPrivSign(EVP_MD_CTX* ctx, EVP_PKEY *pkey, std::string const& msg,
+                std::string& sign, std::string& error) {
+  size_t signLength;
+  if (EVP_DigestSignInit(ctx, nullptr, EVP_sha256(), nullptr, pkey) == 0) {
+    error.append("EVP_DigestSignInit failed: ").append(ERR_error_string(ERR_get_error(), nullptr));
+    return 1;
+  }
+  if (EVP_DigestSignUpdate(ctx, msg.c_str(), msg.size()) == 0) {
+    error.append("EVP_DigestSignUpdate failed: ").append( ERR_error_string(ERR_get_error(), nullptr));
+    return 1;
+  }
+  if (EVP_DigestSignFinal(ctx, nullptr, &signLength) == 0) {
+    error.append("EVP_DigestSignFinal failed: ").append( ERR_error_string(ERR_get_error(), nullptr));
+    return 1;
+  }
+  sign.resize(signLength);
+  if (EVP_DigestSignFinal(ctx, (unsigned char*) sign.data(), &signLength) == 0) {
+    error.append("EVP_DigestSignFinal failed, return code: ")
+      .append(ERR_error_string(ERR_get_error(), nullptr));
+    return 1;
+  }
+  return 0;
+}
+
+int rsaPrivSign(std::string const& pem , std::string const& msg,
+                std::string& sign, std::string& error) {
+
+  BIO* keybio = BIO_new_mem_buf(pem.c_str(), -1);
+  RSA* rsa;
+    rsa = RSA_new();
```

Comment:
Indentation does not look quite right.

Reply:
Furthermore, this can return NULL and then we probably get a crash in the following line.